### ISSUE TITLE
GMMClassifier has overflow in predict_proba [BUG] 

### ISSUE THREAD

**Issue by concepttt-ma** [koaning/scikit-lego#506]
Hey there

Thanks for providing us with the bricks ;)

I was just playing around with the GMMClassifier; use case: classify 300 dimensional word embeddings (that is, i have a set of lexical items per class for which I use the word embeddings as representation and then train the GMMClassifier on it.)

predict() runs fine but predict_proba() stumbled across an overflow:
```
In [129]: gmmclfmod.predict_proba(extended_neg_emb)
/home/ma/.ve/esgdev/lib/python3.8/site-packages/sklego/mixture/gmm_classifier.py:93: RuntimeWarning: overflow encountered in exp
  return np.exp(res) / np.exp(res).sum(axis=1)[:, np.newaxis]
/home/ma/.ve/esgdev/lib/python3.8/site-packages/sklego/mixture/gmm_classifier.py:93: RuntimeWarning: invalid value encountered in true_divide
  return np.exp(res) / np.exp(res).sum(axis=1)[:, np.newaxis]
Out[129]: 
array([[nan,  0.],
       [nan,  0.],
       [nan,  0.],
       [nan,  0.],
       [nan,  0.],
 ...
```
digging a little deeper I found that when I manually grabbed the scores like:
```
In [165]: gmmclfmod.gmms_["POS"].score_samples([extended_pos_emb[0]])
Out[165]: array([2315.93200567])

In [166]: gmmclfmod.gmms_["NEG"].score_samples([extended_pos_emb[0]])
Out[166]: array([-237563.66350931])
```
and then tried to simply do the np.exp() with those, it's already stumbling:
```
In [169]: np.exp([2274.69379128, -349121.15994271])
<ipython-input-169-5d6fb400d75d>:1: RuntimeWarning: overflow encountered in exp
  np.exp([2274.69379128, -349121.15994271])
Out[169]: array([inf,  0.])
```

However, why not using the softmax from scipy (I mean that's what we want to have in the end for the predict_proba, right?)
```
from scipy.special import softmax
In [170]: softmax([2274.69379128, -349121.15994271])
Out[170]: array([1., 0.])
```
-> horray: no problem

In scipy they do the softmax in logspace:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.softmax.html

--> in the source:
```
# compute in log space for numerical stability
 return np.exp(x - logsumexp(x, axis=axis, keepdims=True))
```
So, I don't know if that is a thing you'd just take from scipy or rather want to copy over the codeline; however, this would resolve the bug, I suppose

have a nice day
Michi

**Comment by koaning** [comment]
Are you able to share a bit of the data that might help us reproduce the error? 

Also, if you're interested in doing stuff with word-embeddings ... have you seen [this other library](https://github.com/koaning/whatlies) that I wrote? It's [scikit-learn compatible too](https://koaning.github.io/whatlies/tutorial/scikit-learn/) and comes with some [nice examples](https://koaning.github.io/whatlies/examples/lipstick-pig/).

There are also a bunch of "be careful with sentiment" warnings that I might want to mention as well ([link](https://www.youtube.com/watch?v=0K5ybetv-dA&ab_channel=HuggingFace)).

**Comment by koaning** [comment]
Actually ... I imagine that `from scipy.special import softmax` may be the most elegant solution. 

If you're up for a PR I'd love to receive it. If you do though, you may want to check both the [GMMClassifier](https://github.com/koaning/scikit-lego/blob/main/sklego/mixture/gmm_classifier.py) as well as the [BayesianGMMClassifier](https://github.com/koaning/scikit-lego/blob/main/sklego/mixture/bayesian_gmm_classifier.py). It'd also be nice if we could add a unit-test that confirms the behavior that we're fixing.

**Comment by concepttt-ma** [comment]
> Are you able to share a bit of the data that might help us reproduce the error?
> 
> Also, if you're interested in doing stuff with word-embeddings ... have you seen [this other library](https://github.com/koaning/whatlies) that I wrote? It's [scikit-learn compatible too](https://koaning.github.io/whatlies/tutorial/scikit-learn/) and comes with some [nice examples](https://koaning.github.io/whatlies/examples/lipstick-pig/).
> 
> There are also a bunch of "be careful with sentiment" warnings that I might want to mention as well ([link](https://www.youtube.com/watch?v=0K5ybetv-dA&ab_channel=HuggingFace)).

-----

Sorry for the late reply; I was off on holidays, restricting myself from doing code-related stuff; as I see, you've put that into the code base yourself; thanks. Checked it also on my case, now this works as expected (circumventing the under/overflow problems).

And: yes, I've touched [whatlies](https://github.com/koaning/whatlies) beforehand ... one of the many things-I-could-also-write-a-chapter-with-an-interesting-case-study for my thesis back then ... had a look at the "sentiment-talk" with HF; yes, you are clearly right about those issues, and I personally totally agree that it's not a "solved problem" at all. It just seems that people like it more to have a check sign on the list than dealing with the hard issues, annoying details and ignored problems (well, severe ones) - as long as the solution presents a nine as first digit of the f1 scores ... 

appreciate all your work on bad-labelled data, bias, and important skepticism or awareness concerning available ML solutions

have a great day
Michi
